### DISCUSSION TITLE
How to write complex captions with evaluated R code?

### DISCUSSION THREAD

**Discussion by svraka** [quarto-dev/quarto-cli#1785] [Q&A]
One nice feature of the new YAML-style chunk options is that we can write complex text for e.g. captions without the need for escaping everything in an R character vector (by the way this is not Quarto-specific, it also works with plain knitr).

````
```{r plot1}
#| fig.cap: A caption with some math, $\gamma = 0.5$
boxplot(Sepal.Length ~ Species, data = iris)
```
````

Is there a way to include R code in such a caption? Obviously, we can't use the standard way for evaluating R code by [prefixing the value with `!expr`](https://quarto.org/docs/computations/r.html#chunk-options), as the value in this case is not an R expression. 

In bookdown [text references](https://bookdown.org/yihui/rmarkdown/bookdown-markdown.html#text-references) can be used for such a purpose:

````
```{r plot2}
#| fig.cap = "(ref:plot-cap)"
boxplot(Sepal.Length ~ Species, data = iris)
```

(ref:plot-cap) A caption with some math, $\gamma = 0.5$, $N = `r nrow(iris)`$.
````

Is something like this possible with Quarto without resorting to character vectors?

```{r plot3}
#| fig-cap: !expr sprintf("A caption with some math, $\\gamma = 0.5$, ($N = %s$)", nrow(iris))
boxplot(Sepal.Length ~ Species, data = iris)
```

**Comment by mcanouil** [comment]
I am not sure to understand the question or issue.
The following works just fine and is "complex":
```r
#| fig-cap: !expr sprintf("A caption with some math, $\\gamma = 0.5$, ($N = %s$)", nrow(iris))
boxplot(Sepal.Length ~ Species, data = iris)
```
<img width="1399" alt="image" src="https://user-images.githubusercontent.com/8896044/183962912-b111d184-77d4-4406-a96c-11f474a66fcf.png">

Are you asking for a new feature to be implemented in Quarto, which is the "text reference" thing from bookdown?

**Reply by dmkaplan2000** [reply]
I presume he is asking for a bookdown-like "text references" thing. At the very least, I am asking myself that very same question as text references are a really handy way to develop complex figure and table captions.

**Reply by mcanouil** [reply]
Far as I know and implicitly stated with my last question, it is a feature that does no exist at this point.
So, for now, one have to use regular R expression with !expr prefix

**Reply by svraka** [reply]
Oh, sorry I missed the original reply from @mcanouil. Indeed, my question was about avoiding ugly R expressions with lots of backslashes.

**Reply by mcanouil** [reply]
I let maintainers answer, but this text reference feature does not seem to be really compatible with Jupyter Notebook, thus this would create way more heterogeneity for the users.

To note instead of using LaTeX-style math, you can use HTML entities for symbols, especially if your format is html.

**Reply by dmkaplan2000** [reply]
I use text references for captions that just have a lot of random R code and formatting involved, nothing to do with equations, so this question is more general than any single format.

Perhaps this should be converted into a bug report/feature request?

**Reply by svraka** [reply]
I see, thanks for the explanation!

TBH, HTML entities are even uglier 😆 Although, thinking about a bit more, one could use unicode characters, which can replace a lot of backslashes, and should work for PDF and HTML as well. And glue strings might be more readable, than `sprintf` calls.

````
```{r plot-glue}
#| fig-cap: !expr glue::glue("A caption with some math, $γ = 0.5$, ($N = {nrow(iris)}$)")
boxplot(Sepal.Length ~ Species, data = iris)
```
````

**Reply by mcanouil** [reply]
> I use text references for captions that just have a lot of random R code and formatting involved, nothing to do with equations, so this question is more general than any single format.
> 
> 
> 
> Perhaps this should be converted into a bug report/feature request?

My point was not about the format but about the engine.

At least a feature request, one can create it from this discussion using the link on the menu on the right of the discussion.

**Reply by dmkaplan2000** [reply]
Oops, I went ahead and filed a [feature request](https://github.com/quarto-dev/quarto-cli/issues/1959) before I saw this message...

**Reply by mcanouil** [reply]
so you did not see the "create issue from discussion" button ^^
![image](https://user-images.githubusercontent.com/8896044/185343213-3983bdf0-e764-451a-a66b-45db45efbab2.png)

This automatically links the discussion to the issue, for example from another discussion:
![image](https://user-images.githubusercontent.com/8896044/185343808-e9a77a81-4e28-4d7e-8926-dc185a9fc15f.png)

**Reply by dmkaplan2000** [reply]
No. I see it now, but it is kind of in an out of the way place on my screen. Sorry for the bother...

**Reply by mcanouil** [reply]
It's fine, it's kind of a new feature from GitHub.
I noticed this only few weeks ago. 😉

**Comment by mcanouil** [comment]
For all that might see this discussion: the related issue is #1959 and it is scheduled as part of the [v1.2 release milestone](https://github.com/quarto-dev/quarto-cli/milestone/6).

**Comment by eeholmes** [comment]
I achieve this as follows. My plot captions are often hairy bits of code since I am dynamically creating the caption based on the data being plotted. 

    ```{r}
    plot_cap <-  sprintf("A caption with some math, $\\gamma = 0.5$, ($N = %s$)", nrow(iris))
    ```
or

    ```{r}
    plot_cap <-  captionfunction(mydata)
    ```

and then

    ```{r}
    #| fig-cap: !expr plot_cap
    boxplot(Sepal.Length ~ Species, data = iris)
    ```
  
basically the same thing, but I find this cleaner.

**Reply by egouldo** [reply]
Thanks for the example code. Can the `plot_cap` expression be created within the same chunk where it is called, or does it have to be created in a previous chunk?

**Reply by cderv** [reply]
It can be created in the same code chunk as this is a special **knitr** feature where some chunk options are evaluated **after** chunk evaluation. `fig-cap` (i..e `fig.cap` in **knitr**) is one of them

Did you try it already ? It should work in Quarto for **knitr** engine as this is a knitr feature.

In **knitr**, this is controlled at package option level by `eval.after` knitr option (https://yihui.org/knitr/options/#package-options) and this concerns mainly figure option (`fig.cap`, `fig.scap`, `fig.alt` and their dash version in Quarto).

From knitr documentatin
> `eval.after`: (`c('fig.cap', 'fig.scap', 'fig.alt')`; character) 
> A character vector of option names. These options will be evaluated after a chunk has been evaluated, and all other options will be evaluated before a chunk. For example, for `fig.cap = paste('p-value is', t.test(x)$p.value)`, it will be evaluated after the chunk according to the value of x if `eval.after = 'fig.cap'`.
